index.js

```diff
@@ -19,74 +12,152 @@ function setPrototype(self) {
-
+const siftOptions = { operations: { $where: function() { throw '$WHERE_NOT_ALLOWED'; } } };
+ 
 // Helper functions
-function toDots(d) { return d.replace(/ᐉ/g,'.'); }
 function noop() {}
+function inverseIdentity(d) { return !d; }
+function identity(d) { return d; }
 
 // This is the main prototype 
 var Query = Object.create(Array.prototype);
 
-Query.scale = function(_domain) {
+function astToCluesPath(node) {
+  return astToString(node.paren || node);
+}
+
+function astToString(node) {
+  if (node.piped) {
+    return node.piped.map(n => astToString(n)).join('|');
+  }
+  if (node.equation) {
+    return `${astToString(node.equation.left)}${node.operation}${astToString(node.equation.right)}`;
+  }
+  if (node.paren) {
+    return `(${astToString(node.paren)})`;
+  }
+  if (node.remoteLink) {
+    return `\${${astToString(node.remoteLink)}}`;
+  }
+  if (node.quoted) {
+    return `"${node.quoted.replace(/"/g,'\\"')}"`;
+  }
+  if (node.and) {
+    return `and(${astToString(node.and)})`;
+  }
+  if (node.or) {
+    return `or(${astToString(node.or)})`;
+  }
+  if (node.not) {
+    return `not(${astToString(node.not)})`;
+  }
+  if (Array.isArray(node)) {
+    return node.map(n => astToString(n)).join('.');
+  }
+  return node;
+}
+
+function createExternal(cb) {
+  let counter = 0,
+      context = {};
+
+  return function $external(path) { 
+    let ast = pathParser(path);
+    let firstNode = ast.root;
+    let firstNodeKey = `_counter${astToString(firstNode)}`;
+    let remainder = ast.extra ? `.${ast.extra}` : '';
+
+    let existingCounter = context[firstNodeKey];
+    if (!existingCounter) {
+      existingCounter = ++counter;
+      context[firstNodeKey] = existingCounter;
+      Object.defineProperty(context, `externalKey${existingCounter}`, {value: cb(firstNode), enumerable: true, configurable: true, writable: true});
+    }
+
+    return [context, `externalKey${existingCounter}${remainder}`,identity];
+  };
+}
+
+Query.scale = function(_domain, $global) {
   var self = this;
-  return function $property(key) {
-    key = key.split(reSplitter);
-    var domainKey =  key[1] || _domain;
-    var rangeKey = key[0];
-    if (!domainKey)
+
+  return createExternal(async ast => {
+    let pipePieces = ast.piped ? ast.piped : [ast];
+    if (pipePieces.length >= 3) {
+      throw 'MAXIMUM_OF_THREE_INPUTS';
+    }
+    let rangeKey = astToCluesPath(pipePieces[0]);
+    let domainKey = pipePieces.length === 2 ? astToCluesPath(pipePieces[1]) : _domain;
+    if (!domainKey) {
       throw 'NO_DOMAIN';
-    
+    }
+
+    let domainValues = Promise.map(self, d => clues(d, domainKey, $global).catch(noop));
+    let rangeValues = Promise.map(self, d => clues(d, rangeKey, $global).catch(noop));
+
+    domainValues = await domainValues;
+    rangeValues = await rangeValues;
+
+    // modify the items - but in a single tick, so should be safe
+    self.forEach((d,i) => {
+      d._domainValue = domainValues[i];
```

Comment:
this technically wouldn't work on straight strings or numbers, but `scale` won't work for that either